src/um/devicetopology.rs

```diff
@@ -0,0 +1,487 @@
+STRUCT!{struct KSIDENTIFIER_s {
+    Set: GUID,
+    Id: ULONG,
+    Flags: ULONG,
+}}
+UNION!{union KSIDENTIFIER {
+    [u64; 3],
+    s s_mut: KSIDENTIFIER_s,
+    Alignment Alignment_mut: LONGLONG,
+}}
+pub type KSPROPERTY = KSIDENTIFIER;
+pub type PKSPROPERTY = *mut KSIDENTIFIER;
+pub type KSMETHOD = KSIDENTIFIER;
+pub type PKSMETHOD = *mut KSIDENTIFIER;
+pub type KSEVENT = KSIDENTIFIER;
+pub type PKSEVENT = *mut KSIDENTIFIER;
+ENUM!{enum EPcxConnectionType {
+    eConnTypeUnknown = 0,
+    eConnType3Point5mm = 1,
+    eConnTypeQuarter = 2,
+    eConnTypeAtapiInternal = 3,
+    eConnTypeRCA = 4,
+    eConnTypeOptical = 5,
+    eConnTypeOtherDigital = 6,
+    eConnTypeOtherAnalog = 7,
+    eConnTypeMultichannelAnalogDIN = 8,
+    eConnTypeXlrProfessional = 9,
+    eConnTypeRJ11Modem = 10,
+    eConnTypeCombination = 11,
+}}
+ENUM!{enum EPcxGeoLocation {
+    eGeoLocRear = 1,
+    eGeoLocFront = 2,
+    eGeoLocLeft = 3,
+    eGeoLocRight = 4,
+    eGeoLocTop = 5,
+    eGeoLocBottom = 6,
+    eGeoLocRearPanel = 7,
+    eGeoLocRiser = 8,
+    eGeoLocInsideMobileLid = 9,
+    eGeoLocDrivebay = 10,
+    eGeoLocHDMI = 11,
+    eGeoLocOutsideMobileLid = 12,
+    eGeoLocATAPI = 13,
+    eGeoLocNotApplicable = 14,
+    eGeoLocReserved6 = 15,
+}}
+ENUM!{enum EPcxGenLocation {
+    eGenLocPrimaryBox = 0,
+    eGenLocInternal = 1,
+    eGenLocSeparate = 2,
+    eGenLocOther = 3,
+}}
+ENUM!{enum EPxcPortConnection {
+    ePortConnJack = 0,
+    ePortConnIntegratedDevice = 1,
+    ePortConnBothIntegratedAndJack = 2,
+    ePortConnUnknown = 3,
+}}
+STRUCT!{struct KSJACK_DESCRIPTION {
+    ChannelMapping: DWORD,
+    Color: COLORREF,
+    ConnectionType: EPcxConnectionType,
+    GeoLocation: EPcxGeoLocation,
+    GenLocation: EPcxGenLocation,
+    PortConnection: EPxcPortConnection,
+    IsConnected: BOOL,
+}}
+pub type PKSJACK_DESCRIPTION = *mut KSJACK_DESCRIPTION;
+STRUCT!{struct LUID {
+    LowPart: DWORD,
+    HighPart: LONG,
+}}
+pub type PLUID = *mut LUID;
+ENUM!{enum KSJACK_SINK_CONNECTIONTYPE {
+    KSJACK_SINK_CONNECTIONTYPE_HDMI = 0,
+    KSJACK_SINK_CONNECTIONTYPE_DISPLAYPORT = 1,
+}}
+STRUCT!{struct KSJACK_SINK_INFORMATION {
+    ConnType: KSJACK_SINK_CONNECTIONTYPE,
+    ManufacturerId: WORD,
+    ProductId: WORD,
+    AudioLatency: WORD,
+    HDCPCapable: BOOL,
+    AICapable: BOOL,
+    SinkDescriptionLength: UCHAR,
+    SinkDescription: [WCHAR; 32],
+    PortId: LUID,
+}}
+STRUCT!{struct KSJACK_DESCRIPTION2 {
+    DeviceStateInfo: DWORD,
+    JackCapabilities: DWORD,
+}}
+pub type PKSJACK_DESCRIPTION2 = *mut KSJACK_DESCRIPTION2;
+ENUM!{enum DataFlow {
+    In = 0,
+    Out = 1,
+}}
+ENUM!{enum PartType {
+    Connector = 0,
+    Subunit = 1,
+}}
+ENUM!{enum ConnectorType {
+    Unknown_Connector = 0,
+    Physical_Internal = 1,
+    Physical_External = 2,
+    Software_IO = 3,
+    Software_Fixed = 4,
+    Network = 5,
+}}
+RIDL!(#[uuid(0x28f54685, 0x06fd, 0x11d2, 0xb2, 0x7a, 0x00, 0xa0, 0xc9, 0x22, 0x31, 0x96)]
```

Comment:
Please use `{}` instead of `()` for item level macros such as `RIDL!`.

Reply:
@🐇 Done.